src/test/php/lang/ast/unittest/emit/ReadonlyTest.class.php

```diff
@@ -143,4 +173,23 @@ public function cannot_have_an_initial_value() {
       public readonly string $fixture= "Test";
     }');
   }
+
+  #[Test, Expect(class: Error::class, withMessage: '/Cannot create dynamic property .+fixture/')]
+  public function cannot_read_dynamic_members_from_readonly_classes() {
+    $t= $this->type('readonly class <T> { }');
+    $t->newInstance()->fixture;
+  }
+
+  #[Test, Expect(class: Error::class, withMessage: '/Cannot create dynamic property .+fixture/')]
+  public function cannot_write_dynamic_members_from_readonly_classes() {
+    $t= $this->type('readonly class <T> { }');
+    $t->newInstance()->fixture= true;
+  }
+
+  #[Test, Ignore('Until proper error handling facilities exist')]
+  public function readonly_classes_cannot_have_static_members() {
```

Comment:
This proper error handling means any way of raising an exception from within the emitting phase and having it turned into a lang.ast.Error with file and line set.

Reply:
Maybe these also need to go into the parser - where, instead of returning `lang.ast.Node` instances directly, we'd do something like:

```php
// Before
return new PropertyNode(...);

// After
return $this->ast->property(...);
```

...and have two separate *ast* implementations: checked and unchecked. This would be part of a greater refactoring, thus keeping it for a separate pull request.